lib/AST/Decl.cpp

```diff
@@ -4087,46 +4087,152 @@ bool ProtocolDecl::isAvailableInExistential(const ValueDecl *decl) const {
   return true;
 }
 
-bool ProtocolDecl::existentialTypeSupportedSlow(LazyResolver *resolver) {
+void ProtocolDecl::getAssociatedTypeFreedomMapRec(
+                     AssociatedTypeFreedomMap &map) const {
+  for (auto *proto: getInheritedProtocols())
+    proto->getAssociatedTypeFreedomMapRec(map);
+
+  for (auto *assoc: getAssociatedTypeMembers())
+    map.insert({assoc, false});
+
+  for (auto &req : getRequirementSignature()) {
+    if (req.getKind() == RequirementKind::SameType &&
+        req.getSecondType()->getKind() != TypeKind::DependentMember) {
+
+      if (auto assoc = req.getFirstType()->getAs<DependentMemberType>()) {
+        auto *assocDecl = assoc->getAssocType();
```

Comment:
... the *base* type might not be Self here. For example,

```
protocol Q {
  associatedtype T
  associatedtype U where T : Q, T.T == Int, T.U == Int
}
```

I think your algorithm will flag this protocol as having both associated types fully constrained when they're not.

Instead you should look at the Self archetype of the protocol and iterate over its nested types:

```
auto selfType = protoDecl->getSelfInterfaceType();
auto archetypeType = protoDecl->mapTypeIntoContext(selfType)->castTo<ArchetypeType>()
for (auto nested : archetypeType->getNestedTypes()) {
  ...
}
```

Reply:
> ```swift
> protocol Q {
>  associatedtype T
>  associatedtype U where T : Q, T.T == Int, T.U == Int
> }
> ```
I didn't know associated types accept constraints that aren't related to them. I can't seem to make it compile with this conformance:
```swift
extension Int: Q {
  typealias T = Int
  typealias U = Int
}
```

I believe what you meant here and in the previous [comment](https://github.com/apple/swift/pull/21576#discussion_r244833860) is dealt with right after by checking whether `req.getFirstType()` corresponds to an associated type declaration *present* in the map:
```cpp
 if (map.count(assocDecl))
   map[assocDecl] = true;
```
Somehow `Assoc.Assoc` is always a different `AssociatedTypeDecl` than `Assoc`. For example, this works. 
```swift
protocol P { associatedtype Q: P }
protocol PP: P where Q.Q == Int {}

func foo2(arg: PP) {} // protocol 'PP' can only be used ...
```
Or... is `getAssocType()` just giving me a different reference? But all other cases work fine. Could it be giving me a different reference only when it's a nested associated type?